cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go

```diff
@@ -61,13 +61,13 @@ func encodeTokenSecretData(token *BootstrapToken, now time.Time) map[string][]by
 	if token.Expires != nil {
 		// Format the expiration date accordingly
 		// TODO: This maybe should be a helper function in bootstraputil?
-		expirationString := token.Expires.Time.Format(time.RFC3339)
+		expirationString := token.Expires.Time.UTC().Format(time.RFC3339)
```

Comment:
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/cluster-bootstrap/util/secrets/secrets.go#L64

It will be used here and it looks like the changes should be safe. The scope of the impact is just the Annotation record time for UTC

Reply:
would like ack from cluster-lifecycle reviewer ...

/assign @neolit123

Reply:
thank you very much

Reply:
the parsing from cluster-bootstrap seems fine:
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/cluster-bootstrap/util/secrets/secrets.go#L64

but if generated tokens client side start using UTC, my assumption is that server side the token cleaner and bootstrap signer controllers should be adapted as well.
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/controller/bootstrap/tokencleaner.go#L192

https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go#L131

if we make the change server side, to me this seems like we are introducing a weird scenario for tokens that already exist in etcd.

this was mentioned here in the other closed PR:
https://github.com/kubernetes/kubernetes/pull/102451#discussion_r642546852

i could be wrong.

Reply:
> the parsing from cluster-bootstrap seems fine:
> https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/cluster-bootstrap/util/secrets/secrets.go#L64
> 
> but if generated tokens client side start using UTC, my assumption is that server side the token cleaner and bootstrap signer controllers should be adapted as well.
> https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/controller/bootstrap/tokencleaner.go#L192
> 
> https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go#L131
> 
> if we make the change server side, to me this seems like we are introducing a weird scenario for tokens that already exist in etcd.
> 
> this was mentioned here in the other closed PR:
> [#102451 (comment)](https://github.com/kubernetes/kubernetes/pull/102451#discussion_r642546852)
> 
> i could be wrong.

Thank you very much, do you mean that the time zone used in the secrets already stored in etcd is local, but after the update, the subsequent stored secrets are changed to utc. may cause confusion to the user?

Reply:
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/cluster-bootstrap/util/secrets/secrets.go#L64

the tokencleaner should not need to be modified, it uses "time.Parse" parsing and will parse to UTC time zone

Reply:
the controllers currently accept local time and `time.Now() != time.Now().UTC()`

Reply:
> the controllers currently accept local server time and `time.Now() != time.Now().UTC()`

Sorry, I didn't understand

`time.Now().Equal(time.Now().UTC())` is true

Reply:
i think i'm ignoring the fact that when a time object is marshaled to RFC3339 for local time it stores the time zone offset.
which means that parsing the string back to a time object would result in the same time.

so starting to store time as UTC strings in boostrap objects doesn't seem like a breaking change.
it is just cosmetic.

/approve

Reply:
> i think i'm ignoring the fact that when a time object is marshaled to RFC3339 for local time it stores the time zone offset.
> which means that parsing the string back to a time object would result in the same time.
> 
> so starting to store time as UTC strings in boostrap token objects doesn't seem like a breaking change.
> it is just cosmetic.
> 
> /approve

As you said, thanks very much!